File: Reservation/src/main/java/net/travel/reservation/services/PaiementService.java
package net.travel.reservation.services;

import lombok.RequiredArgsConstructor;
import net.travel.reservation.dto.PaiementRequestDTO;
import net.travel.reservation.entites.Paiement;
import net.travel.reservation.entites.Reservation;
import net.travel.reservation.repositories.PaiementRepository;
import net.travel.reservation.repositories.ReservationRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class PaiementService {

    private final PaiementRepository paiementRepository;
    private final ReservationRepository reservationRepository;

    public List<Paiement> getAllPaiements() {
        return paiementRepository.findAll();
    }

    public Paiement getPaiementById(UUID id) {
        return paiementRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Paiement introuvable avec l'ID : " + id));
    }

    @Transactional
    public Paiement createPaiement(PaiementRequestDTO dto) {
        if (dto.getMontant() == null || dto.getMontant().compareTo(BigDecimal.ZERO) <= 0) {
            throw new RuntimeException("Le montant du paiement doit être supérieur à zéro");
        }

        if (dto.getReservationId() == null) {
            throw new RuntimeException("Le paiement doit être associé à une réservation valide");
        }

        Reservation reservation = reservationRepository.findById(dto.getReservationId())
                .orElseThrow(() -> new RuntimeException("Le paiement doit être associé à une réservation valide"));

        BigDecimal totalDejaPaye = calculerTotalPaye(reservation.getReservationId());
        BigDecimal nouveauTotal = totalDejaPaye.add(dto.getMontant());

        if (reservation.getMontantAPayer() != null && nouveauTotal.compareTo(reservation.getMontantAPayer()) > 0) {
            throw new RuntimeException("Le montant total payé (" + nouveauTotal +
                    ") dépasse le montant total de la réservation (" + reservation.getMontantAPayer() + ")");
        }

        Paiement paiement = Paiement.builder()
                .montant(dto.getMontant())
                .datePaiement(dto.getDatePaiement())
                .typePaiement(dto.getTypePaiement())
                .methodePaiement(dto.getMethodePaiement())
                .reservation(reservation)
                .notes(dto.getNotes())
                .build();

        return paiementRepository.save(paiement);
    }

    @Transactional
    public Paiement updatePaiement(UUID id, Paiement paiementRequest) {
        Paiement paiement = getPaiementById(id);

        if (paiementRequest.getMontant() != null && paiementRequest.getMontant().compareTo(BigDecimal.ZERO) <= 0) {
            throw new RuntimeException("Le montant du paiement doit être supérieur à zéro");
        }

        paiement.setMontant(paiementRequest.getMontant());
        paiement.setTypePaiement(paiementRequest.getTypePaiement());
        paiement.setMethodePaiement(paiementRequest.getMethodePaiement());
        paiement.setNotes(paiementRequest.getNotes());
        paiement.setUrlRecuPdf(paiementRequest.getUrlRecuPdf());

        return paiementRepository.save(paiement);
    }

    @Transactional
    public void deletePaiement(UUID id) {
        Paiement paiement = getPaiementById(id);
        paiementRepository.delete(paiement);
    }

    public List<Paiement> getPaiementsByReservation(Long reservationId) {
        return paiementRepository.findByReservationReservationId(reservationId);
    }

    public BigDecimal calculerTotalPaye(Long reservationId) {
        List<Paiement> paiements = getPaiementsByReservation(reservationId);
        return paiements.stream()
                .map(Paiement::getMontant)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
}

File: reservation-frontend/src/app/core/services/paiement.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Paiement } from '../models/paiement/paiement';
import { PaiementRequestDTO } from '../models/PaiementRequestDTO';
@Injectable({
  providedIn: 'root'
})
export class PaiementService {

  private apiUrl = 'http://localhost:8099/api/paiements';

  constructor(private http: HttpClient) { }

  // Tous les paiements
  getAllPaiements(): Observable<Paiement[]> {
    return this.http.get<Paiement[]>(this.apiUrl);
  }

  // Paiement par ID
  getPaiementById(id: string): Observable<Paiement> {
    return this.http.get<Paiement>(`${this.apiUrl}/${id}`);
  }

  // Ajouter (avec DTO)
  createPaiement(paiement: PaiementRequestDTO): Observable<Paiement> {
    return this.http.post<Paiement>(this.apiUrl, paiement);
  }

  // Modifier
  updatePaiement(id: string, paiement: Paiement): Observable<Paiement> {
    return this.http.put<Paiement>(`${this.apiUrl}/${id}`, paiement);
  }

  // Supprimer
  deletePaiement(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Paiements d'une réservation
  getPaiementsByReservation(reservationId: number): Observable<Paiement[]> {
    return this.http.get<Paiement[]>(
      `${this.apiUrl}/reservation/${reservationId}`
    );
  }

  // Total payé d'une réservation
  calculerTotalPaye(reservationId: number): Observable<number> {
    return this.http.get<number>(
      `${this.apiUrl}/reservation/${reservationId}/total`
    );
  }
}